src/paho/mqtt/client.py

```diff
@@ -805,27 +826,27 @@ def reconnect(self):
             if err.errno != errno.EINPROGRESS and err.errno != errno.EWOULDBLOCK and err.errno != EAGAIN:
                 raise
 
-        if self._tls_ca_certs is not None:
-            sock = ssl.wrap_socket(
-                sock,
-                certfile=self._tls_certfile,
-                keyfile=self._tls_keyfile,
-                ca_certs=self._tls_ca_certs,
-                cert_reqs=self._tls_cert_reqs,
-                ssl_version=self._tls_version,
-                ciphers=self._tls_ciphers)
-
-            if self._tls_insecure is False:
-                if sys.version_info < (2,7,9) or (sys.version_info[0] == 3 and sys.version_info[1] < 2):
-                    self._tls_match_hostname(sock)
-                else:
-                    ssl.match_hostname(sock.getpeercert(), self._host)
+        if self._ssl:
+            # SSL is only supported when SSLContext is available (implies Python >= 2.7.9 or >= 3.2)
 
-        if self._transport == "websockets":
-            if self._tls_ca_certs is not None:
-                sock = WebsocketWrapper(sock, self._host, self._port, True)
+            verify_host = not self._tls_insecure
+            try:
+                # Try with server_hostname, even it's not supported in certain scenarios
+                sock = self._ssl_context.wrap_socket(sock, server_hostname=self._host)
+            except ValueError:
```

Comment:
Should be fixed now.

Looking at the ssl package docs, ssl.CertificateError should be the only error raised by the ssl code that isn't derived from OSError.